Update dependencies

ralpha opened the pull request:
## Info
This PR changes:
* Bump minimum supported Rust version from `1.41.0` to `1.45.0`
* Update `lsp-types` from `0.82` to `0.89`
* Update `tokio` from `0.2` to `1.4.0`
* Update `tokio-util` from `0.3` to `0.6.5`
* Update `bytes` from `0.5` to `1.0.1`
* Update `dashmap` from `3.5.1` to `4.0.2`
* Update `nom` from `5.1` to `6.1.2`
* Update `env_logger` from `0.7` to `0.8.3` (dev-dependencies)
* Update `tower-test` from `0.3` to `0.4` (dev-dependencies)
This PR does not add any new features.

Because of changes in dependencies this version needs to be a minor version update, so `0.14.0`.
These changes where tested (and succeeded) on:
* `cargo test --all` (1.51.0, 2021-03-23)
* `cargo +nightly test --all` (1.53.0, 2021-03-24 )
* `cargo build` (1.51.0, 2021-03-23)
* `cargo +nightly build` (1.53.0, 2021-03-24 )
* tested examples
* tested on my project using VSCode

This PR has effect on other PR's and issues:
* Closes: #263, #262, #254, #253, #252, #250, #249, #234
* Effects: #257, #243

### Extra info
Clippy warning was not fixed so only version updates and necessary changes are in PR.
```
warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
  --> tower-lsp-macros/src/lib.rs:48:1
   |
48 | fn parse_method_calls<'a>(lang_server_trait: &'a ItemTrait) -> Vec<MethodCall<'a>> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::needless_lifetimes)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
```

Crate checked with:
* `cargo +nightly udeps` => "All deps seem to have been used."
* `cargo audit` => "Success No vulnerable packages found"
* `cargo outdated` => "All dependencies are up to date, yay!" (2021-04-08)

ralpha commented:
@ebkalderon Could you please merge this PR and publish a new version on crates.io. This way the repo is kept maintained and up to date for a bit longer, thanks.

ebkalderon reviewed (approved):
Sorry for the delayed review, @ralpha. These changes look excellent! Thank you. :rocket:

Just need to rebase on top of the latest `master` and it should be good to merge. I can open a separate PR to fix the `clippy` lint as well, if necessary.

ralpha commented:
I updated the PR: 
- Upstream changes from `master` are now merged.
- Added `Cargo.lock` as this should be included in the repo to make builds reproducible.
- Fixed 1 linting error where lifetime was not needed, there are more linting warnings but they require more changes. These should be done in an other PR.
- Updated tokio to 1.6.0 ( and thus Closes #268 )
- Some other minor dependencies where also updated (patches) as they are now checked by `Cargo.lock`

ralpha commented:
Hopefully there can be a new release after this (and maybe other) PR's are merged so I can resolve other dependence conflicts in my repo downstream.

ebkalderon commented:
Thanks for rebasing and updating the PR! Yeah, I'll be pushing a new release as soon as the backlog of other pulls requests is properly tended to.

One small matter to note, though, is that standard Rust library crates tend to usually not check in the `Cargo.lock` into version control. This is an intentional decision on the part of the Cargo developers. See https://github.com/rust-lang/cargo/issues/315 for the original reasoning behind this.

ralpha commented:
> One small matter to note, though, is that standard Rust library crates tend to usually not check in the `Cargo.lock` into version control. This is an intentional decision on the part of the Cargo developers. See [rust-lang/cargo#315](https://github.com/rust-lang/cargo/issues/315) for the original reasoning behind this.

Yeah you are right, did not notice/know this before, learned something :)
It is also in the docs: https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html and https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

Removed the commit/lock file from the repo and history. Other changes are the same (except for lock file changes)

ebkalderon commented:
No worries! Thanks so much for the pack of updates, and especially for updating the `CHANGELOG` as part of the PR. 🚀
